### PR TITLE
Spring boot 2 compatibility and move to prometheus

### PR DESCRIPTION
**Pull request by mrandi**
Added compatibility with spring boot, and start using prometheus.

For compatibility we would like to move to the prometheus endpoint instead of implement the old one by ourself.

### CODE REVIEW

**Comment by mrandi** [comment]
What do you think about?

**Comment by Jan-M** [comment]
Looks good to me, supporting new spring boot versions.

Just not sure why when using a vendor neutral api suddenly "prometheus" has to pop up in the API path?

**Comment by mrandi** [comment]
@Jan-M a few points to support it :)
- The first is that this were already implemented https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-metrics.html#production-ready-metrics and to spare some time we though to use them 
- we already use prometheus 
- we would like to use the implementation provided by spring in order to stay up to date also with newer versions (no implementation nor maintenance on our side needed)

File: pom.xml [hunk 1]
```diff
@@ -1,28 +1,22 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
     <parent>
-        <groupId>org.zalando.stups.build</groupId>
-        <artifactId>stups-parent</artifactId>
-        <version>6</version>
-        <relativePath />
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
```
Comment (harti2006) [inline pom.xml:5]: please keep in mind, that changing this parent pom will remove the Maven Central deploy configuration. It needs to be added manually here in the pom again.

File: samples/zmon-actuator-jetty-jersey/src/test/java/com/zalando/zmon/boot/jetty/BaseJerseyConfigIT.java [hunk 2]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright (C) 2015 Zalando SE (http://tech.zalando.com)
- *
+ * <p>
```
Comment (harti2006) [inline samples/zmon-actuator-jetty-jersey/src/test/java/com/zalando/zmon/boot/jetty/BaseJerseyConfigIT.java:3]: What about removing license headers instead reformatting them ;-)
If I remember correctly they are discouraged by the OpenSource guidelines

File: zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonMetricsFilterTest.java [hunk 3]
```diff
@@ -60,31 +49,32 @@
...
 
     private final Random random = new Random(System.currentTimeMillis());
 
     @Before
     public void setUp() {
-        ConsoleReporter reporter = ConsoleReporter.forRegistry(metricRegistry).convertRatesTo(TimeUnit.SECONDS)
-                                                  .convertDurationsTo(TimeUnit.MILLISECONDS).build();
-        reporter.start(2, TimeUnit.SECONDS);
-
+//        Reporter reporter = Reporter.forRegistry(meterRegistry).convertRatesTo(TimeUnit.SECONDS)
```
Comment (harti2006) [inline zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonMetricsFilterTest.java:60]: TODO: find a replacement for the console logger or remove the commented code

**Comment by harti2006** [comment]
Please note that Prometheus was only used in the test to expose the collected metrics via API. The former `/metrics` endpoint has gone (or basically was replaced with completely different semantics).

As Michele said: Using Prometheus for the actual ZMON integration (into the Cloud Dashboard) would be our suggestion, because ZMON worker can already handle the Prometheus format. From what we can see in the [docs](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#production-ready-metrics-export), all other other natively supported metrics exporters follow a "push model", where data is actively sent from the Spring Boot app to e.g. Datadog, Atlas, etc. Only Prometheus supports a passive mode and exposes metrics via API in the application. JMX would be another alternative.

**Comment by harti2006** [comment]
BTW, this is how the old `/metrics` endpoint looks now with Spring Boot 2.0:
```
curl -s http://127.0.0.1:8080/actuator/metrics/ | jq '.'
{
  "names": [
    "jvm.memory.max",
    "http.server.requests",
    "process.files.max",
    "jvm.gc.memory.promoted",
    "tomcat.cache.hit",
    "system.load.average.1m",
    "tomcat.cache.access",
    "jvm.memory.used",
    "jvm.gc.max.data.size",
    "jvm.gc.pause",
    "jvm.memory.committed",
    "system.cpu.count",
    "logback.events",
    "tomcat.global.sent",
    "jvm.buffer.memory.used",
    "tomcat.sessions.created",
    "jvm.threads.daemon",
    "system.cpu.usage",
    "jvm.gc.memory.allocated",
    "tomcat.global.request.max",
    "tomcat.global.request",
    "tomcat.sessions.expired",
    "jvm.threads.live",
    "jvm.threads.peak",
    "tomcat.global.received",
    "process.uptime",
    "tomcat.sessions.rejected",
    "process.cpu.usage",
    "tomcat.threads.config.max",
    "jvm.classes.loaded",
    "jvm.classes.unloaded",
    "tomcat.global.error",
    "tomcat.sessions.active.current",
    "tomcat.sessions.alive.max",
    "jvm.gc.live.data.size",
    "tomcat.servlet.request.max",
    "tomcat.threads.current",
    "tomcat.servlet.request",
    "process.files.open",
    "jvm.buffer.count",
    "jvm.buffer.total.capacity",
    "tomcat.sessions.active.max",
    "tomcat.threads.busy",
    "process.start.time",
    "tomcat.servlet.error"
  ]
}
```

You only receive a list of available metric keys. Then you can go on and query each key, like:
```
curl -s http://127.0.0.1:8080/actuator/metrics/jvm.memory.max | jq '.'
{
  "name": "jvm.memory.max",
  "measurements": [
    {
      "statistic": "VALUE",
      "value": 5600444415
    }
  ],
  "availableTags": [
    {
      "tag": "area",
      "values": [
        "heap",
        "nonheap"
      ]
    },
    {
      "tag": "id",
      "values": [
        "Compressed Class Space",
        "PS Survivor Space",
        "PS Old Gen",
        "Metaspace",
        "PS Eden Space",
        "Code Cache"
      ]
    }
  ]
}
```

As an extra feature, you can drill down using tags:

```
curl -s http://127.0.0.1:8080/actuator/metrics/jvm.memory.max\?tag\=area:nonheap | jq '.' 
{
  "name": "jvm.memory.max",
  "measurements": [
    {
      "statistic": "VALUE",
      "value": 1325400063
    }
  ],
  "availableTags": [
    {
      "tag": "id",
      "values": [
        "Metaspace",
        "Compressed Class Space",
        "Code Cache"
      ]
    }
  ]
}
```

**Comment by mohabusama** [comment]
@harti2006 So this update would introduce breaking changes to current monitoring?

**Comment by mrandi** [comment]
@mohabusama exactly, with spring 2 we have breaking changes. We still have the possibility to adopt the changes based on what we had before but I don’t believe that this will make sense.

**Comment by duergner** [comment]
I'm basically in favour of this PR the only thing I don't like that is mixing new functionality and simple reformatting. Makes it rather hard to properly review and understand which parts are just reformatting and which parts are actually introducing something new.

I'd suggest to first to the reformatting PR and afterwards have a smaller one which does introduce the actual changes.

File: zmon-actuator/pom.xml [hunk 4]
```diff
@@ -1,83 +1,93 @@
...
+                <artifactId>jacoco-maven-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>org.eluder.coveralls</groupId>
+                <artifactId>coveralls-maven-plugin</artifactId>
+                <version>3.1.0</version>
+                <configuration>
+                    <dryRun>${coveralls.dryRun}</dryRun>
+                    <jacocoReports>
+                        <jacocoReport>target/site/jacoco/jacoco.xml</jacocoReport>
```
Comment (whiskeysierra) [inline zmon-actuator/pom.xml:87]: I don't believe this is needed.

File: zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonRestResponseBackendMetricsInterceptorTest.java [hunk 5]
```diff
@@ -15,47 +15,49 @@
...
-                }
-            });
+                });
 
         ArgumentCaptor<StopWatch> stopwatchArgumentCaptor = ArgumentCaptor.forClass(StopWatch.class);
         interceptor.intercept(null, null, execution);
 
-        Mockito.verify(mockedWrapper, times(1)).recordBackendRoundTripMetrics(Mockito.any(HttpRequest.class),
-            Mockito.any(ClientHttpResponse.class), stopwatchArgumentCaptor.capture());
+        verify(mockedWrapper, times(1))
```
Comment (whiskeysierra) [inline zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonRestResponseBackendMetricsInterceptorTest.java:57]: `times(1)` is the default.

File: pom.xml (hunk 1 above)
Comment (whiskeysierra) [reply pom.xml:5]: Also, this is a breaking change, since we will no longer support Spring Boot 1.x. The new version should bump the major version.

**Comment by Jan-M** [comment]
All in favor of bumping major version for this, should not become a breaking change for existing systems. Especially if we aim for new Spring Boot version anyways.

**Comment by harti2006** [comment]
I prepared another PR #24 that does all the code formatting and some minor dependency upgrades. Idea is to rebase Spring Boot 2 upgrade later on top of those changes.

**Comment by MALPI** [comment]
Can we make some progress here?

**Comment by MALPI** [comment]
I think this can be closed in favor of #26. I will try to apply the comments that have already been made.

File: pom.xml (hunk 1 above)
Comment (MALPI) [reply pom.xml:5]: Applied in #26

File: samples/zmon-actuator-jetty-jersey/src/test/java/com/zalando/zmon/boot/jetty/BaseJerseyConfigIT.java (hunk 2 above)
Comment (MALPI) [reply samples/zmon-actuator-jetty-jersey/src/test/java/com/zalando/zmon/boot/jetty/BaseJerseyConfigIT.java:3]: Has been done on master

File: zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonMetricsFilterTest.java (hunk 3 above)
Comment (MALPI) [reply zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonMetricsFilterTest.java:60]: I dropped the code.

File: zmon-actuator/pom.xml (hunk 4 above)
Comment (MALPI) [reply zmon-actuator/pom.xml:87]: Should I drop only jacoco or what do you mean?

File: zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonRestResponseBackendMetricsInterceptorTest.java (hunk 5 above)
Comment (MALPI) [reply zmon-actuator/src/test/java/org/zalando/zmon/actuator/ZmonRestResponseBackendMetricsInterceptorTest.java:57]: Changed

**Pull request closed without merging**